### PR TITLE
middleware to fire telemetry events

### PR DESCRIPTION
**Pull request by hez**
This allows you to subscribe to the telemetry events and collect metrics about your jobs. We figured this might be useful to other people in the Exq community.

We used something like the following to collect and display using telemetry metrics:
```
defmodule MyApp.Application do
  def start(_type, _args) do
    children = [
      # .....
      {:telemetry_poller, measurements: periodic_measurements(), period: 10_000},
      {Telemetry.Metrics.ConsoleReporter, metrics: metrics()}
    ]

    opts = [strategy: :one_for_one, name: MyApp.Supervisor]
    Supervisor.start_link(children, opts)
  end

  defp metrics do
    [
      counter("exq.job.started.count"),
      counter("exq.job.started.retry_count"),
      counter("exq.job.processed.retry_count"),
      distribution("exq.job.processed.duration",
        buckets: [0.1, 0.2, 0.3, 0.5, 0.75, 1, 2, 3, 5, 10]
      ),
      counter("exq.job.failed.retry_count"),
      distribution("exq.job.failed.duration",
        buckets: [0.1, 0.2, 0.3, 0.5, 0.75, 1, 2, 3, 5, 10]
      )
    ]
  end
end
```

Co-authored-by: Gregory Daynes <gregdaynes@gmail.com>

### CODE REVIEW

File: lib/exq/middleware/telemetry_middleware.ex [hunk 1]
```diff
@@ -0,0 +1,86 @@
+defmodule Exq.TelemetryMiddleware do
```
[the remark this answers is not in the document]
Comment (hez) [reply lib/exq/middleware/telemetry_middleware.ex:1]: Good catch, missed that when moving it from our repo.

File: lib/exq/middleware/telemetry_middleware.ex [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    )
+
+    pipeline
+  end
+
+  def before_work(pipeline) do
+    job = pipeline.assigns.job
+
+    :telemetry.execute(
+      [:exq, :job, :started],
```
[the remark this answers is not in the document]
Comment (hez) [reply lib/exq/middleware/telemetry_middleware.ex:72]: Definitely could, I had originally used the same names as this library used. Will make it `start`, `exception`, `stop` 👍

File: lib/exq/middleware/telemetry_middleware.ex [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+          buckets: [0.1, 0.2, 0.3, 0.5, 0.75, 1, 2, 3, 5, 10]
+        ),
+        counter("exq.job.failed.retry_count"),
+        distribution("exq.job.failed.duration",
+          buckets: [0.1, 0.2, 0.3, 0.5, 0.75, 1, 2, 3, 5, 10]
+        )
+      ]
+    end
+  end
+  ```
```
[the remark this answers is not in the document]
Comment (hez) [reply lib/exq/middleware/telemetry_middleware.ex:34]: Good call

File: lib/exq/middleware/telemetry.ex [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    )
+
+    assign(pipeline, :started_at, DateTime.utc_now())
+  end
+
+  defp tags(job),
+    do: %{enqueued_at: unix_to_datetime(job.enqueued_at), queue: job.queue, class: job.class}
+
+  defp delta(%Pipeline{assigns: assigns}),
+    do: DateTime.diff(DateTime.utc_now(), assigns.started_at, :second)
```
Comment (hez) [inline lib/exq/middleware/telemetry.ex:100]: @ananthakumaran How do you feel about this for calculating the delta? I would rather use the built in DateTime diff function then manually calculate it like in the Logger middleware, not sure if that breaks your minimum ver of Elixir though.

Comment (ananthakumaran) [reply lib/exq/middleware/telemetry.ex:100]: there are two things
1) duration should be calculated based on monotonic time, so we can just do a normal subtraction, the unit would be native [reference](https://github.com/whatyouhide/redix/blob/a249060f761ce85f87ad91cfab1dec74ee45aeb2/lib/redix.ex#L786)
2) for the start event, there will also a system_time field calculated using `System.system_time()`. This in addition with duration could be used to calculate the span time if required.

**Comment by ananthakumaran** [comment]
@hez I have made some changes to make it the events same as telemetry:span/3. Let me know if you have any comments. I will try to merge it to master in a week or so.

**Comment by ananthakumaran** [comment]
@akira could you cut a new release from the current master. I am working on a couple of features/bug fixes that I would like to merge into master once a release is made from the current master.

**Comment by hez** [comment]
@ananthakumaran Thank you for getting those fixes in there! Sorry things have been hectic here and this was lower on the todo list :\

**Comment by akira** [comment]
Sure sounds good - I will cut a release on this at the latest this  coming
weekend.

Thanks!

**Pull request merged**